universal-editor-product-picker/src/universal-editor-ui-1/web-src/src/components/ProductPicker.js

```diff
@@ -0,0 +1,302 @@
+      Object.values(categories).forEach(c => {
+        c.key = `category:${c.id}`;
+        c.isFolder = true;
+      });
+
+      setState(state => {
+        return {
+          ...state,
+          categories,
+        };
+      });
+    })();
+  }, []);
+
+  // breadcrumbs
+  useEffect(() => {
+    (async () => {
+      let breadcrumbs = [];
+      if (state.categories.length !== 0 && state.searchText.length === 0) {
+        breadcrumbs = state.categories[state.currentCategory]
+          .path.split('/')
+          .map(p => state.categories[p])
+          .filter(p => p);
+      }
+
+      setState(state => {
+        return {
+          ...state,
+          breadcrumbs,
+        };
+      });
+    })();
+  }, [state.categories, state.currentCategory, state.searchText]);
+
+  // items: categories + 1st page of products
+  useEffect(() => {
+    setState(state => {
+      return {
+        ...state,
+        loadingState: 'loading',
+      };
+    });
+
+    (async () => {
+      let categories = [];
+      if (state.categories.length !== 0 && state.searchText.length === 0) {
+        categories = Object.values(state.categories)
+          .filter(category => category.parentId === state.currentCategory);
+      }
+
+      let products = {};
+      let pageInfo = {};
+      try {
+        [products, pageInfo] = await getProducts(
+          config['commerce-endpoint'], 
+          state.currentCategory, 1, 
+          state.searchText
+        );
+      } catch (err) {
+        setState(state => ({
+          ...state,
+          error: 'Could not load items',
+        }));
+        return;
+      }
+      Object.values(products).forEach(i => {
+        i.key = i.sku;
+      });
+
+      const items = [...categories, ...Object.values(products)];
+
+      setState(state => {
+        return {
+          ...state,
+          loadingState: 'idle',
+          items,
+          pageInfo,
+        };
+      });
+    })();
+  }, [state.categories, state.currentCategory, state.searchText]);
+
+  // @TODO
+  // selected products
+  // useEffect(() => {
+  //   setState(state => ({
+  //     ...state,
+  //     selectedProducts: selectedProducts || [],
+  //   }));
+  // }, [selectedProducts]);
+
+  // only for products, categories are always displayed as a complete list
+  const onLoadMore = async () => {
+    if (state.pageInfo.current_page >= state.pageInfo.total_pages || state.loadingState === 'loadingMore') {
+      return;
+    }
+
+    setState(state => ({
+      ...state,
+      loadingState: 'loadingMore',
+    }));
+
+    const [products, pageInfo] = await getProducts(
+      config['commerce-endpoint'],
+      state.currentCategory, 
+      state.pageInfo.current_page + 1, 
+      state.searchText
+    );
+    Object.values(products).forEach(i => {
+      i.key = i.sku;
+    });
+    const newItems = [ ...state.items, ...Object.values(products) ];
+
+    setState(state => {
+      return {
+        ...state,
```

Comment:
No need to act on it, but for future: I think this type of complex state management can be effectively  done using useReducer() pattern of react hooks.